File: apps/elixir/test/algorithm/sort_algorithms_test.exs
defmodule Algorithm.SortAlgorithmsTest do
  use ExUnit.Case
  alias Algorithm.SortAlgorithms

  @sorted [1, 2, 3, 4, 5, 6, 7, 8]
  @unsorted [5, 3, 8, 1, 4, 7, 2, 6]

  test "bubble_sort/1" do
    assert SortAlgorithms.bubble_sort(@unsorted) == @sorted
  end

  test "selection_sort/1" do
    assert SortAlgorithms.selection_sort(@unsorted) == @sorted
  end

  test "insertion_sort/1" do
    assert SortAlgorithms.insertion_sort(@unsorted) == @sorted
  end

  test "shell_sort/1" do
    assert SortAlgorithms.shell_sort(@unsorted) == @sorted
  end

  test "quick_sort/1" do
    assert SortAlgorithms.quick_sort(@unsorted) == @sorted
  end

  test "merge_sort/1" do
    assert SortAlgorithms.merge_sort(@unsorted) == @sorted
  end

  test "heap_sort/1" do
    assert SortAlgorithms.heap_sort(@unsorted) == @sorted
  end

  test "counting_sort/2" do
    assert SortAlgorithms.counting_sort([3, 1, 4, 1, 5, 9, 2, 6], 9) ==
             [1, 1, 2, 3, 4, 5, 6, 9]
  end
end

File: apps/scala/src/test/scala/algorithm/SortAlgorithmsTest.scala
package algorithm

import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

class SortAlgorithmsTest extends AnyFunSuite with Matchers:

  val unsorted  = Array(5, 3, 8, 1, 9, 2, 7, 4, 6)
  val ascending = Array(1, 2, 3, 4, 5, 6, 7, 8, 9)

  test("bubbleSort"):
    val a = unsorted.clone()
    SortAlgorithms.bubbleSort(a)
    a shouldBe ascending

  test("selectionSort"):
    val a = unsorted.clone()
    SortAlgorithms.selectionSort(a)
    a shouldBe ascending

  test("insertionSort"):
    val a = unsorted.clone()
    SortAlgorithms.insertionSort(a)
    a shouldBe ascending

  test("shellSort"):
    val a = unsorted.clone()
    SortAlgorithms.shellSort(a)
    a shouldBe ascending

  test("quickSort"):
    val a = unsorted.clone()
    SortAlgorithms.quickSort(a)
    a shouldBe ascending

  test("mergeSort"):
    SortAlgorithms.mergeSort(unsorted) shouldBe ascending

  test("heapSort"):
    val a = unsorted.clone()
    SortAlgorithms.heapSort(a)
    a shouldBe ascending

  test("countingSort"):
    SortAlgorithms.countingSort(unsorted) shouldBe ascending

  test("ソート: 空配列"):
    SortAlgorithms.bubbleSort(Array.empty[Int])
    SortAlgorithms.selectionSort(Array.empty[Int])
    SortAlgorithms.insertionSort(Array.empty[Int])
    SortAlgorithms.mergeSort(Array.empty[Int]) shouldBe Array.empty[Int]
    SortAlgorithms.countingSort(Array.empty[Int]) shouldBe Array.empty[Int]

  test("ソート: 要素が 1 つ"):
    val a = Array(42)
    SortAlgorithms.bubbleSort(a)
    a shouldBe Array(42)

  test("ソート: 重複あり"):
    val a = Array(3, 1, 4, 1, 5, 9, 2, 6, 5)
    SortAlgorithms.bubbleSort(a)
    a shouldBe Array(1, 1, 2, 3, 4, 5, 5, 6, 9)
